services/web/server/src/simcore_service_webserver/socketio/messages.py

```diff
@@ -32,37 +30,80 @@
 SOCKET_IO_WALLET_OSPARC_CREDITS_UPDATED_EVENT: Final[str] = "walletOsparcCreditsUpdated"
 
 
+async def _safe_emit(
+    sio: AsyncServer,
+    *,
+    room: SocketIORoomStr,
+    message: SocketMessageDict,
+    ignore_queue: bool,
+):
+    # NOTE 1 : we configured message queue (i.e. socketio servers are backed with rabbitMQ)
+    # so if `ignore_queue=True` then the server can directly communicate with the
+    # client without having to send his message first to rabbitMQ and then back to itself.
+    #
+    # NOTE 2: `emit` method is not designed to be used concurrently
+    try:
+        event = message["event_type"]
+        data = json_dumps(message["data"])
+        await sio.emit(
+            event=event,
+            data=data,
+            room=room,
+            ignore_queue=ignore_queue,
+        )
+    except Exception:  # pylint: disable=broad-exception-caught
+        _logger.warning(
+            "Failed to deliver %s message to %s size=%d",
+            f"{event=}",
+            f"{room=}",
+            len(data),
+            exc_info=True,
+        )
+
+
 async def send_messages_to_user(
-    app: Application, user_id: UserID, messages: Sequence[SocketMessageDict]
+    app: Application,
+    user_id: UserID,
+    message: SocketMessageDict,
+    *,
+    has_direct_connection_to_client: bool = True,
 ) -> None:
+    """
+    Keyword Arguments:
+        has_direct_connection_to_client -- set to False when this message is delivered from a server that has no direct connection to the client (default: {True})
+        An example where this is value is False, is sending messages to a user in the GC
+
+    QUESTION: a user might have different tabs opened or connect with a different browser/computer. Does stickiness
+    make all these connection associated to the same server?
```

Comment:
stickiness works IP based. so different tabs will go to the same server where a different computer might not.
why is this a question?